### PR TITLE
DAOS-14010 rebuild: add delay rebuild

### PR DESCRIPTION
**Pull request by wangdi1**
Add "delay rebuild" healing mode, so the delay rebuild process is

1) SWIM detects dead ranks and report to the PS leader, which update
the pool map, i.e. marking the related targets as DOWN.
2) Though the rebuild job will not be scheduled, until there are further
manual pool operations, for example drain, extend, reintegration.
3) Then all these pool operations will be merged into one rebuild job,
then scheduled.

Update placement algothrim to be able to calculate the layout with
merged pool operation.

Reply rebuild version to client side to make sure inflight IO will not
be sent to those not-rebuilding nodes, for example if some nodes are
only just being added to the pool map, though not being reintegrated
yet, then the inflight I/O needs to ignore these nodes.

Abort the rebuild job immediately if it finds further pool map update,
so the current job will be merged to the following rebuild job. So
concurrent pool operation will be allowed, no EBUSY check anymore.

Add various tests to verify the delay rebuild process.

Features: rebuild
Required-githooks: true
Remove pool operation EBUSY check.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by daltonbohning** [comment]
> requested review from mjmac (assigned from daos-stack/go-watchers), tanabarr (assigned from daos-stack/go-watchers), liuxuezhao and jolivier23 and removed request for daos-stack/go-watchers, daos-stack/ftest-watchers, daos-stack/client-api-watchers, daos-stack/vos-watchers, mjmac and tanabarr

Please do not remove the auto-added watchers. That defeats the purpose

**Review by daltonbohning — DISMISSED** [review]
This needs to run with
`Features: rebuild`
At least the modified test `src/tests/ftest/control/log_entry.py` is not a `pr` test, and I suspect other rebuild tests would be affected by these changes as well.

**Review by mjmac — COMMENTED** [review]
Go changes LGTM.

Probably out of scope for this change, but we may want to add some more metrics to provide insight into the state/progress of rebuilds so that this information can be easily added to dashboards instead of having to query the pool.

**Comment by wangdi1** [comment]
> Go changes LGTM.
> 
> Probably out of scope for this change, but we may want to add some more metrics to provide insight into the state/progress of rebuilds so that this information can be easily added to dashboards instead of having to query the pool.

sure, will add some metrics for rebuild status.

File: src/tests/ftest/daos_test/suite.yaml
Comment (wangdi1) [reply src/tests/ftest/daos_test/suite.yaml:25]: I thought this is only for failure case, to check if the test really timeout. Normally, it should not take this long.

Comment (daltonbohning) [reply src/tests/ftest/daos_test/suite.yaml]: There isn't a different timeout for failure vs success. If the test shouldn't take this long, then increasing the timeout shouldn't help

**Comment by jolivier23** [comment]
looks like you have a conflict now, probably due to your put patch?

**Comment by wangdi1** [comment]
> looks like you have a conflict now, probably due to your put patch?

oh, yes, let me refresh the patch.

**Comment by wangdi1** [comment]
failure is unrelated. ping

**Comment by wangdi1** [comment]
failure due to DAOS-14598

**Comment by wangdi1** [comment]
ping reviewers. thanks

File: src/tests/ftest/daos_test/suite.yaml
Comment (wangdi1) [reply src/tests/ftest/daos_test/suite.yaml]: rebuild process changed a bit. in some nasty environment, it may take longer here.

Comment (daltonbohning) [reply src/tests/ftest/daos_test/suite.yaml]: Even in the CI environment? For a PR test, 1 hour is a long time

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by jolivier23 — DISMISSED** [review]

File: src/common/pool_map.c
Comment (wangdi1) [reply src/common/pool_map.c]: Sure, I will update if it needs to refresh. Thanks!

**Comment by wangdi1** [comment]
@liuxuezhao @gnailzenh  Could you please review? Thanks

File: src/rebuild/scan.c
Comment (liuxuezhao) [inline src/rebuild/scan.c:932]: is this comment still needed?
the rank used in pool map is rank in primary group, so seems no problem for the usage.

File: src/rebuild/srv.c
Comment (liuxuezhao) [inline src/rebuild/srv.c:648]: just confirm why started the rebuild (with the rgt) with "dom->do_comp.co_fseq > rgt->rgt_rebuild_ver" (and later abort it here)? the rebuild_prepare() should not start it?

File: src/vos/vos_obj_index.c
Comment (liuxuezhao) [inline src/vos/vos_obj_index.c:244]: seems no usage?

**Review by liuxuezhao — COMMENTED** [review]
to confirm some details, thx

File: src/pool/srv_pool_map.c
Comment (wangdi1) [reply src/pool/srv_pool_map.c:31]: ah, this part should be removed, this is used to revert the failure extend rebuild, which has been moved to another function. this should be removed. Thanks.

File: src/common/pool_map.c
Comment (wangdi1) [reply src/common/pool_map.c:1936]: yeah, sure

File: src/common/pool_map.c
Comment (wangdi1) [reply src/common/pool_map.c:1876]: I think you are right.

File: src/common/pool_map.c
Comment (wangdi1) [reply src/common/pool_map.c:1914]: sure, let's follow the original rule

File: src/object/srv_obj.c
Comment (wangdi1) [reply src/object/srv_obj.c:2313]: because with delay rebuild patch, the pool map can be changed during other pool rebuilding, so if there are new extending/reintegrate happened, the pool map might change, and broadcast to the client, then client use the new pool map to generate the layout, but if the extend/reint/rebuild are not started on the server yet. Then it may corrupt the data.

So we use this check to let the inflight I/O wait until the correspondent rebuild start.

File: src/object/srv_obj_migrate.c
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:3106]: in the original code, we do not chose the target randomly for reintegrate operation, but since we merged all operations together, so let's remove those random chosing for the moment.

File: src/placement/jump_map.c
Comment (wangdi1) [reply src/placement/jump_map.c:199]: These are to make sure those delay rebuild(DOWN2UP) targets are being picked to be rebuilt as well. ok, I will add some comments here.

File: src/placement/jump_map_versions.c
Comment (wangdi1) [reply src/placement/jump_map_versions.c:30]: correct.

File: src/placement/jump_map_versions.c
Comment (wangdi1) [reply src/placement/jump_map_versions.c:21]: NEW means the pool map does not know these target yet. so they can be ignored for layout generation.

All other states needs to be considered in jump hash, and those DOWN/DOWNOUT targets will remapped during layout generation,  otherwise it will cause data corruption.

is_comp_avaible means if the component is available for the layout, i.e. whether the shard can choose this target, no need remap.

is_excluded_comp() means if the component needs to be considered in this layout generation (jump hash process). so they are totally different.

File: src/placement/pl_map_common.c
Comment (wangdi1) [reply src/placement/pl_map_common.c:252]: oh, the target is originally as UPIN, then turns to DOWN, but due to delay rebuild (with DOWN2UP flag), then it turns to UP without rebuild, so the original status should be UPIN.

Note: this is pre_rebuild layout generation, basically, it calculate the layout by pervious "stable" version, and post_rebuild is to calculate the layout by post "stable" version(after rebuild),  then find the diff to know which shard needs to be rebuilt.

File: src/placement/pl_map_common.c
Comment (wangdi1) [reply src/placement/pl_map_common.c:264]: ah, right. thanks.

File: src/placement/pl_map_common.c
Comment (wangdi1) [reply src/placement/pl_map_common.c:272]: oh, new status is excluded from layout generation process process anyway.

File: src/rebuild/scan.c
Comment (wangdi1) [reply src/rebuild/scan.c:932]: ok, I will remove it

File: src/rebuild/srv.c
Comment (wangdi1) [reply src/rebuild/srv.c:648]: this means the pool map has been changed(co_fseq) during the rebuild, so let's abort the current rebuild and resolve them altogether in the next rebuild, since we have delay-rebuild, which has the ability to merge all operations in one rebuild.  Or I miss your question?

File: src/vos/vos_obj_index.c
Comment (wangdi1) [reply src/vos/vos_obj_index.c:244]: ah, forget remove it.

File: src/placement/jump_map.c
Comment (liuxuezhao) [reply src/placement/jump_map.c:202]: the "reint_tgt == original_target" can be removed.

**Review by liuxuezhao — DISMISSED** [review]

**Review by daltonbohning — DISMISSED** [review]
Test tags aren't carried forward, so the latest commit would probably need
```
Features: rebuild
```

File: src/control/lib/daos/pool_property.go
Comment (mjmac) [inline src/control/lib/daos/pool_property.go:73]: Please add accompanying unit test cases for this new value: https://github.com/daos-stack/daos/blob/master/src/control/lib/daos/pool_property_test.go#L177

File: src/control/server/config/server_legacy.go
Comment (mjmac) [inline src/control/server/config/server_legacy.go:2]: Why were these copyright dates updated?

File: src/control/lib/daos/pool_property.go
Comment (wangdi1) [reply src/control/lib/daos/pool_property.go:73]: sure, thanks

File: src/control/server/config/server_legacy.go
Comment (wangdi1) [reply src/control/server/config/server_legacy.go]: ah, will revert this.

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
ping

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by jolivier23 — APPROVED** [review]

**Review by kjacque — APPROVED** [review]
Go changes LGTM, although seems like a number of copyright updates in otherwise unchanged files. I wouldn't block on that though.

**Comment by wangdi1** [comment]
Failure is due to this IOR warning, size query has some issues.  I do not think this is related.

2024-02-05 13:56:54,231 process          L0686 INFO | Command '/usr/bin/dmg -o /etc/daos/daos_control.yml -d system stop --force --ranks=11' finished with 0 after 2.9162449836730957s
2024-02-05 13:56:54,279 command_utils    L1332 INFO | Updating the expected state for rank 11 on wolf-205: joined -> ['stopped', 'excluded']
2024-02-05 13:57:42,853 process          L0416 DEBUG| [stdout] WARNING: Expected aggregate file size       = 274877906944
2024-02-05 13:57:42,853 process          L0416 DEBUG| [stdout] WARNING: Stat() of aggregate file size      = 275882770432
2024-02-05 13:57:42,853 process          L0416 DEBUG| [stdout] WARNING: Using actual aggregate bytes moved = 274877906944
2024-02-05 13:57:42,853 process          L0416 DEBUG| [stdout] write     3382.47    422.90     0.075669    8388608    8192       2.01       77.48      0.000227   77.50      0   
2024-02-05 13:57:42,853 process          L0416 DEBUG| [stdout] Verifying contents of the file(s) just written.

**Comment by wangdi1** [comment]
Just to be clear, I compared all these failures with weekly run of #366, no new issue here. i.e. all failures in my run are exactly same as what failed in weekly run #366.  I already mentioned ticket number in the above comments.  
So @daos-stack/daos-gatekeeper  Could you please considering to land this PR? it has been 3 months. Btw: this should not be 2.8 features, instead of it is 2.6 features, which fixed a lot EBUSY failure in our CI tests.  Thanks.

**Comment by wangdi1** [comment]
After I revert 13571 PR on master, most of failures are "fixed" except 15007, which is known issue about "aggregation is not observed" , absolutely unrelated with this PR. 

https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-13761/2/

So @daos-stack/iof-gatekeeper  Could you please land this PR? thanks.

**Comment by daltonbohning** [comment]
@wangdi1 Should we land the revert (#13761) first, then merge master into this PR and let testing run? That would verify this PR doesn't introduce further regressions on the already-failing tests.
I.e. if the tests are already failing, then we don't know if this PR would have passed them or not.

That said, I think @gnailzenh was okay with landing this one, so I will defer.

**Comment by jolivier23** [comment]
> @wangdi1 Should we land the revert (#13761) first, then merge master into this PR and let testing run? That would verify this PR doesn't introduce further regressions on the already-failing tests. I.e. if the tests are already failing, then we don't know if this PR would have passed them or not.
> 
> That said, I think @gnailzenh was okay with landing this one, so I will defer.

I think it's simpler in the long run to land it and fix #13761 but to me it depends on the type of failures.  If they are failing PRs, we should revert.  If not, I think it's better to fix.

**Comment by wangdi1** [comment]
> @wangdi1 Should we land the revert (#13761) first, then merge master into this PR and let testing run? That would verify this PR doesn't introduce further regressions on the already-failing tests. I.e. if the tests are already failing, then we don't know if this PR would have passed them or not.
> 
> That said, I think @gnailzenh was okay with landing this one, so I will defer.

oh, all failures only happened in weekly run. not in PR. So I do not think 13761 affect any landing for the moment.  Maybe it is easier to wait for fan yong to fix it as @jolivier23  said.

**Comment by wangdi1** [comment]
failure caused by DAOS-15007, 15135 and 15127.  all known failures. ping @daos-stack/daos-gatekeeper

**Review by gnailzenh — APPROVED** [review]

File: src/object/srv_obj_migrate.c
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:3107]: because the PR will merge all the rebuild request into one operation, so it wont't be able to tell if it is rebuilt or reintegrate, since all exclude targets or reintegrate targets will be done all together anyway.

File: src/vos/vos_internal.h
Comment (Nasf-Fan) [inline src/vos/vos_internal.h:363]: Sorry, where are these new members used except being assigned as some value? I do not see where they are returned back to upper layer in the patch.

Comment (wangdi1) [reply src/vos/vos_internal.h:363]: ah, they will used by incremental reintegration, and this is for delay rebuild only. So they will be used later once incremental reintegration is landed, https://github.com/daos-stack/daos/pull/13358.  Though this is targeted for 2.8

Comment (wangdi1) [reply src/vos/vos_internal.h:363]: https://github.com/daos-stack/daos/pull/13879  to refine the issue

**Pull request merged**
